File: cli/src/ipc.ts
import { connect } from 'node:net'
import type { JsonRpcRequest } from '@edge-sync/protocol'

/**
 * 对内核 IPC（Unix socket 行分隔 JSON-RPC）发起单次调用。
 * 一次调用一条连接；同步类方法（task.trigger）在内核侧阻塞到完成，
 * 默认超时给足大文件同步的余量。
 */
export async function ipcCall<T = unknown>(
  socketPath: string,
  method: string,
  params: Record<string, unknown> = {},
  timeoutMs = 30 * 60 * 1000,
): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const socket = connect(socketPath)
    let buf = ''
    let done = false

    const finish = (fn: () => void) => {
      if (done) return
      done = true
      clearTimeout(timer)
      socket.destroy()
      fn()
    }

    const timer = setTimeout(() => {
      finish(() => reject(new Error(`IPC ${method}: timed out after ${timeoutMs}ms`)))
    }, timeoutMs)

    socket.on('connect', () => {
      const req: JsonRpcRequest = { jsonrpc: '2.0', id: 1, method: method as never, params }
      socket.write(JSON.stringify(req) + '\n')
    })

    socket.on('data', (chunk) => {
      buf += chunk.toString('utf8')
      const idx = buf.indexOf('\n')
      if (idx < 0) return
      const line = buf.slice(0, idx)
      finish(() => {
        let resp: { result?: T; error?: { code: number; message: string } }
        try {
          resp = JSON.parse(line)
        } catch (e) {
          reject(new Error(`IPC ${method}: bad response line: ${(e as Error).message}`))
          return
        }
        if (resp.error) {
          reject(new Error(resp.error.message))
          return
        }
        resolve(resp.result as T)
      })
    })

    socket.on('error', (e) => finish(() => reject(e)))
  })
}

File: cmd/edge-sync/ipc.go
// IPC 行协议与 YAML 读写（CLI 侧）。
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"time"

	"gopkg.in/yaml.v3"

	"edge-sync/pkg/protocol"
)

// ---------- IPC（unix socket 行分隔 JSON-RPC） ----------

type ipcError struct {
	Code int
	Msg  string
}

func (e *ipcError) Error() string { return e.Msg }

func dialSocket(path string) (net.Conn, error) {
	conn, err := net.Dial("unix", path)
	if err != nil {
		return nil, fmt.Errorf("无法连接内核（edge-syncd 未运行？）: %w", err)
	}
	// 同步触发最长等待 30 分钟（大文件同步）。
	_ = conn.SetDeadline(time.Now().Add(30 * time.Minute))
	return conn, nil
}

func roundTrip(conn net.Conn, method string, params map[string]any, result any) error {
	rawParams, err := json.Marshal(params)
	if err != nil {
		return err
	}
	req, err := json.Marshal(protocol.Request{
		JSONRPC: "2.0",
		ID:      1,
		Method:  method,
		Params:  rawParams,
	})
	if err != nil {
		return err
	}
	if _, err := conn.Write(append(req, '\n')); err != nil {
		return fmt.Errorf("发送请求失败: %w", err)
	}
	line, err := bufio.NewReader(conn).ReadString('\n')
	if err != nil {
		return fmt.Errorf("读取响应失败: %w", err)
	}
	var resp protocol.Response
	if err := json.Unmarshal([]byte(line), &resp); err != nil {
		return fmt.Errorf("响应解析失败: %w", err)
	}
	if resp.Error != nil {
		return &ipcError{Code: resp.Error.Code, Msg: resp.Error.Message}
	}
	if result != nil && len(resp.Result) > 0 {
		if err := json.Unmarshal(resp.Result, result); err != nil {
			return fmt.Errorf("结果解析失败: %w", err)
		}
	}
	return nil
}

// ---------- YAML ----------

func readYAML(path string) (map[string]any, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("读取配置失败（-c 指定路径）: %w", err)
	}
	var data map[string]any
	if err := yaml.Unmarshal(raw, &data); err != nil {
		return nil, fmt.Errorf("解析 %s: %w", path, err)
	}
	return data, nil
}

func writeYAML(path string, data map[string]any) error {
	raw, err := yaml.Marshal(data)
	if err != nil {
		return err
	}
	return os.WriteFile(path, raw, 0o644)
}
